vmm/src/vm.rs

```diff
@@ -390,24 +416,73 @@ impl DeviceManager {
         vm_fd: &Arc<VmFd>,
         vm_cfg: &VmConfig,
         msi_capable: bool,
+        userspace_ioapic: bool,
     ) -> DeviceManagerResult<Self> {
         let io_bus = devices::Bus::new();
         let mut mmio_bus = devices::Bus::new();
 
-        // Add serial device
-        let serial_evt = EventFd::new(EFD_NONBLOCK).map_err(DeviceManagerError::EventFd)?;
-        let serial_irq_fd = serial_evt.as_raw_fd();
-        let serial_irq_cb =
-            Box::new(move || serial_evt.write(1)) as devices::legacy::SerialInterrupt;
+        let ioapic = if userspace_ioapic {
+            let vm_fd_clone = vm_fd.clone();
+            let ioapic_cb = Box::new(move |msg: ioapic::MsiMessage| {
+                let msi = kvm_msi {
+                    address_lo: msg.addr,
+                    address_hi: 0x0,
+                    data: msg.data,
+                    flags: 0u32,
+                    devid: 0u32,
+                    pad: [0u8; 12],
+                };
+
+                match vm_fd_clone.signal_msi(msi) {
```

Comment:
Here we rely on signal_msi() is always available. Depends on how modern the KVM we want to support we may need either handle it or fail gracefully at an earlier time.

Reply:
This is already checked from the highest level possible in the code: https://github.com/intel/cloud-hypervisor/blob/master/vmm/src/vm.rs#L769

If KVM reports the capability as not being supported, then we forward the information the VMM won't be MSI capable, and instead it will use pin based interrupts.

Reply:
And FWIW, I checked that KVM_SIGNAL_MSI was available in kernel 3.14, so I don't think there's anything to worry about since we don't expect cloud-hypervisor to run on old kernel like that.

Reply:
> This is already checked from the highest level possible in the code: https://github.com/intel/cloud-hypervisor/blob/master/vmm/src/vm.rs#L769
> 

From what I read from the code and from this current PR, we will be using a userspace IOAPIC when both TSC_DEADLINE_TIMER and SPLIT_IRQ are one. Then the userspace IOAPIC relies on MSI, independently of CAP_MSI. I may be missing something as this code is getting complex...

Reply:
Oh yes you're right, I should get the MSI_CAP value first and then decide to use the userspace IOAPIC based on it too.

Reply:
Done!